### ISSUE TITLE
Recursive getter does not throw warning when using pattern matching

### ISSUE THREAD

**Issue by victor-tinoco** [dart-lang/linter#4913]
So, if you declare a class with a getter, like this:

```dart
class A {
  String get name {
    return switch (this) {
      // Notice that while declaring the name getter, I can access it here without warning.
      A(name: final name) => name,
    };
  }
}
```

The `recursive_getters` rule should warn this case, shouldn't it? Pattern matching does allow you to refer to the getter you're declaring (on its declaration scope body). The same occurs when you create a getter through an extension. 

I fell in this case because of my copilot, when I tried to build it by running the tests, it just crashed with a non-descriptive error, probably because it is like a recursion with no break. It should not be allowed since it always falls in a _runtime_ error.

_Obs.: Re-opening it from https://github.com/dart-lang/language/issues/3652._

**Comment by lrhn** [comment]
It's not always a mistake to invoke the same getter, of its bout on the same object.
Fx:
```dart
  Node get last { 
    return switch (this.next) {
      // Notice that while declaring the name getter, I can access it here without warning.
      A(last: final last) => last,
      _ => null,
    };
  }
```
Same getter called, but not on the same object.

So this issue requires not just figuring out that the same getter is office again, but also determining whether it's definitely, or possibly, on the same object.
That's not always going to be possible.

**Comment by victor-tinoco** [comment]
> It's not always a mistake to invoke the same getter, of its bout on the same object. Fx:
> 
> ```dart
>   Node get last { 
>     return switch (this.next) {
>       // Notice that while declaring the name getter, I can access it here without warning.
>       A(last: final last) => last,
>       _ => null,
>     };
>   }
> ```
> 
> Same getter called, but not on the same object.
> 
> So this issue requires not just figuring out that the same getter is office again, but also determining whether it's definitely, or possibly, on the same object. That's not always going to be possible.

Yea, we might apply this only when using switch for `this`, for example. I mean, we have a linter with exactly this proposal `recursive_getters`, it looks like it surely might cover it.
